central/group/datastore/datastore_impl.go

```diff
@@ -294,41 +295,46 @@ func (ds *dataStoreImpl) validateAndPrepGroupForUpsertNoLock(ctx context.Context
 		return errox.AlreadyExists.Newf("cannot update group to default group of auth provider %q as a default group already exists",
 			newGroup.GetProps().GetAuthProviderId())
 	}
-	if err := ds.verifyReferencedRoleAndProvider(newGroup); err != nil {
+	if err := ds.verifyReferencedRoleAndProvider(newGroup, authProviderRegistry); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (ds *dataStoreImpl) verifyReferencedRoleAndProvider(group *storage.Group) error {
+func (ds *dataStoreImpl) verifyReferencedRoleAndProvider(group *storage.Group, authProviderRegistry authproviders.Registry) error {
 	role, found, err := ds.roleDatastore.GetRole(datastoresAccessCtx, group.GetRoleName())
 	if err != nil {
 		return err
 	}
+	groupID := group.GetProps().GetId()
 	if !found {
-		return errox.InvalidArgs.Newf("group %q role name %q does not exist", group.GetProps().GetId(), group.GetRoleName())
+		return errox.InvalidArgs.Newf("group %q role name %q does not exist", groupID, group.GetRoleName())
 	}
-	if err := declarativeconfig.VerifyReferencedResourceOrigin(role, group.GetProps(), role.GetName(), group.GetProps().GetId()); err != nil {
+	if err := declarativeconfig.VerifyReferencedResourceOrigin(role, group.GetProps(), role.GetName(), groupID); err != nil {
 		return err
 	}
 
-	authProvider, found, err := ds.authProviderDatastore.GetAuthProvider(datastoresAccessCtx, group.GetProps().GetAuthProviderId())
-	if err != nil {
-		return err
+	authProviderID := group.GetProps().GetAuthProviderId()
+	authProvider := authProviderRegistry.GetProvider(authProviderID)
```

Comment:
This is where the main change from the auth provider datastore to auth provider in-memory registry happens. This should avoid race conditions where an auth provider is already present in the registry and `/v1/authproviders`, but not present in the database (hence failing the group creation).